src/NLog/LogEventInfo.cs

```diff
@@ -523,6 +536,31 @@ private void CalcFormattedMessage()
             }
         }
 
+        private bool ResetFormattedProperties()
+        {
+            if (this.properties != null && this.properties.Count > 0 && this.logMessageFormatter.HasProperties(this))
+            {
+                for (int i = 0; i < this.parameters.Length; ++i)
+                {
+                    object parameterName = null;
+                    foreach (var property in this.properties)
```

Comment:
The TemplateParser could probably generate a "Hole"-dictionary (Wraps a readonly array). And this "Hole"-dicationary, then has the ability to internally allocate a real dictionary if wanting to add/override properties. Then "just" have to make an artificial enumerator.